src/WebJobs.Script/Description/FunctionGenerator.cs

```diff
@@ -110,31 +119,47 @@ public static Type Generate(string functionAssemblyName, string typeName, Collec
                 il.Emit(OpCodes.Ldloc_1);
                 il.Emit(OpCodes.Ldloc_0);
                 il.Emit(OpCodes.Callvirt, invokeMethod);
-
-                // Copy back out and ref parameters
-                for (int i = 0; i < function.Parameters.Count; i++)
+                
+                
+                if (function.Parameters.Any(p => p.Type.IsByRef))
                 {
-                    var param = function.Parameters[i];
-                    if (!param.Type.IsByRef)
+                    // Store the result into a local Task
+                    il.Emit(OpCodes.Stloc_2);
+
+                    // Load the result Task
+                    il.Emit(OpCodes.Ldloc_2);
+
+                    // We need to wait on the function's task if we have any out/ref
+                    // parameters to ensure they have been populated before we copy them back
+                    il.Emit(OpCodes.Callvirt, typeof(Task).GetMethod("Wait", Type.EmptyTypes));
+
+                    // Copy back out and ref parameters
+                    for (int i = 0; i < function.Parameters.Count; i++)
                     {
-                        continue;
-                    }
+                        var param = function.Parameters[i];
+                        if (!param.Type.IsByRef)
+                        {
+                            continue;
+                        }
 
-                    il.Emit(OpCodes.Ldarg, i);
+                        il.Emit(OpCodes.Ldarg, i);
 
-                    il.Emit(OpCodes.Ldloc_0);
-                    il.Emit(OpCodes.Ldc_I4, i);
-                    il.Emit(OpCodes.Ldelem_Ref);
-                    il.Emit(OpCodes.Castclass, param.Type.GetElementType());
+                        il.Emit(OpCodes.Ldloc_0);
+                        il.Emit(OpCodes.Ldc_I4, i);
+                        il.Emit(OpCodes.Ldelem_Ref);
+                        il.Emit(OpCodes.Castclass, param.Type.GetElementType());
+
+                        il.Emit(OpCodes.Stind_Ref);
+                    }
 
-                    il.Emit(OpCodes.Stind_Ref);
+                    il.Emit(OpCodes.Ldloc_2);
```

Comment:
Good point. I was following the pattern we use with the other locals and remember you had mentioned this before. I'll update this.